src/internal/connector/exchange/service_iterators.go

```diff
@@ -81,84 +32,110 @@ func FilterContainersAndFillCollections(
-				err,
-				errs)
-
-			if qp.FailFast {
-				return errs
-			}
+			errs = support.WrapAndAppend(qp.ResourceOwner, err, errs)
+			continue
 		}
 
 		edc := NewCollection(
 			qp.ResourceOwner,
 			dirPath,
-			collectionType,
+			oi,
 			service,
 			statusUpdater,
 		)
-		collections[*c.GetId()] = &edc
+		collections[cID] = &edc
 
 		fetchFunc, err := getFetchIDFunc(qp.Category)
 		if err != nil {
-			errs = support.WrapAndAppend(
-				qp.ResourceOwner,
-				err,
-				errs)
-
-			if qp.FailFast {
-				return errs
-			}
-
+			errs = support.WrapAndAppend(qp.ResourceOwner, err, errs)
 			continue
 		}
 
-		dirID := *c.GetId()
-		oldDelta := oldDeltas[dirID]
-
-		jobs, delta, err := fetchFunc(ctx, edc.service, qp.ResourceOwner, dirID, oldDelta)
+		jobs, delta, err := fetchFunc(ctx, edc.service, qp.ResourceOwner, cID, oldDeltas[cID])
 		if err != nil {
-			errs = support.WrapAndAppend(
-				qp.ResourceOwner,
-				err,
-				errs,
-			)
+			errs = support.WrapAndAppend(qp.ResourceOwner, err, errs)
 		}
 
 		edc.jobs = append(edc.jobs, jobs...)
 
 		if len(delta) > 0 {
-			deltaURLs[dirID] = delta
+			deltaURLs[cID] = delta
+		}
+
+		// add the current path for the container ID to be used in the next backup
+		// as the "previous path", for reference in case of a rename or relocation.
+		if itp, err := resolver.IDToPath(ctx, cID); err != nil {
+			errs = support.WrapAndAppend(qp.ResourceOwner, err, errs)
+		} else {
+			// Some root folders have no display name. The exchange path will
+			// fail without that, so we inject a fallback here just for
+			// the previousPaths handling.
+			els := itp.Elements()
+			if len(els) == 0 || len(els[len(els)-1]) == 0 {
+				itp = itp.Append(rootFolderAlias)
+			}
+
+			p, err := itp.ToDataLayerExchangePathForCategory(
+				qp.Credentials.AzureTenantID,
+				qp.ResourceOwner,
+				qp.Category,
+				false)
+			if err != nil {
+				errs = support.WrapAndAppend(qp.ResourceOwner, err, errs)
+			} else {
+				prevPaths[cID] = p.Folder()
+			}
 		}
 	}
 
-	col, err := makeMetadataCollection(
+	if col, err := graph.MakeMetadataCollection(
 		qp.Credentials.AzureTenantID,
 		qp.ResourceOwner,
+		graph.DeltaTokenFileName,
+		path.ExchangeService,
 		qp.Category,
 		deltaURLs,
 		statusUpdater,
-	)
-	if err != nil {
-		errs = support.WrapAndAppend("making metadata collection", err, errs)
+	); err != nil {
+		errs = support.WrapAndAppend("making delta metadata collection", err, errs)
+	} else if col != nil {
+		collections[graph.DeltaMetadataCollectionKey] = col
+	}
+
+	if col, err := graph.MakeMetadataCollection(
+		qp.Credentials.AzureTenantID,
+		qp.ResourceOwner,
+		graph.PreviousPathFileName,
+		path.ExchangeService,
+		qp.Category,
+		prevPaths,
+		statusUpdater,
+	); err != nil {
+		errs = support.WrapAndAppend("making paths metadata collection", err, errs)
 	} else if col != nil {
-		collections[metadataKey] = col
+		collections[graph.PathsMetadataCollectionKey] = col
```

Comment:
this is going to cause undefined behavior in KW when it tries to consume it. We now have two collections that have the same `FullPath()`. We should instead be creating a single metadata collection with two files in it

Reply:
Bother.  I knew something about that was tickling my brain.  Should be fixed now.